sys/riscv/riscv/cheri_revoke_machdep.c

```diff
@@ -0,0 +1,513 @@
+vm_do_cheri_revoke(int *res, const struct vm_cheri_revoke_cookie *crc,
+    const uint8_t * __capability crshadow, vm_cheri_revoke_test_fn ctp,
+    uintcap_t * __capability cutp, uintcap_t cut, vm_offset_t start,
+    vm_offset_t end)
+{
+	int perms = cheri_getperm(cut);
+	CHERI_REVOKE_STATS_FOR(crst, crc);
+
+	if (perms == 0) {
+		/* For revoked or permissionless caps, do nothing. */
+
+		/*
+		 * XXX technically, a sealed permissionless thing is a
+		 * bearer token, but that's not really something we use.  We
+		 * should probably insist that people don't; keep one sw bit
+		 * set or something.
+		 */
+
+		CHERI_REVOKE_STATS_BUMP(crst, caps_found_revoked);
+	} else if (cheri_gettag(cut) && ctp(crshadow, cut, perms, start, end)) {
+		void * __capability cscratch;
+		int ok;
+
+		uintcap_t cutr = cheri_revoke_cap(cut);
+
+		CHERI_REVOKE_STATS_BUMP(crst, caps_found);
+
+		/*
+		 * Load-link the position under test; verify that it matches
+		 * our previous load; store conditionally the revoked
+		 * version back.  If the verification fails, don't try
+		 * anything fancy, just modify the return value to flag the
+		 * page as dirty.
+		 *
+		 * It's possible that this CAS will fail because the pointer
+		 * has changed during our test.  That's fine, if this is not
+		 * a stop-the-world scan; we'll catch it in the next go
+		 * around.  However, because CAS can fail for reasons other
+		 * than an actual data failure, we might loop.
+		 *
+		 * Because revoked capabilities are still tagged, one might
+		 * worry that this would reset the capdirty bits.  That's
+		 * not true, tho', because we're storing via the direct
+		 * mapping of physical memory.
+		 */
+again:
+		__asm__ __volatile__ (
+			"lr.c.cap %[cscratch], (%[cutp])\n\t"
+			"cseqx %[ok], %[cscratch], %[cut]\n\t"
+			"beq x0, %[ok], 1f\n\t"
+			"sc.c.cap %[cutr], (%[cutp])\n\t"
+			"1:\n\t"
+		  : [ok] "=r" (ok), [cscratch] "=&C" (cscratch),
+		    [cutr] "+C" (cutr)
+		  : [cut] "C" (cut), [cutp] "C" (cutp)
+		  : "memory");
+
+		/* sc.c.cap clobbers its value operand with 0 on success */
+		if (__builtin_expect((uintcap_t)cutr == 0, 1)) {
+			CHERI_REVOKE_STATS_BUMP(crst, caps_cleared);
+			/* Don't count a revoked cap as HASCAPS */
+		} else if (!cheri_gettag(cscratch)) {
+			/* Data; don't sweat it */
+		} else if (cheri_revoke_is_revoked(cscratch)) {
+			/* Revoked cap; don't worry about it */
+		} else if (__builtin_expect(ok, 1)) {
+			/* Spurious CAS failure */
+			goto again;
+		} else {
+			/* An unexpected capability */
+			*res |= VM_CHERI_REVOKE_PAGE_DIRTY
+				| VM_CHERI_REVOKE_PAGE_HASCAPS ;
+		}
+	} else {
+		CHERI_REVOKE_STATS_BUMP(crst, caps_found);
+
+		/*
+		 * Even though it might actually be un-tagged, that's a very
+		 * narrow race and this a very common case, so don't bother
+		 * testing.  We'll find it clear next time, maybe.
+		 */
+		*res |= VM_CHERI_REVOKE_PAGE_HASCAPS;
+	}
+
+	return (0);
+}
+
+static inline void
+enable_user_memory_access()
```

Comment:
These make me so sad.  Probably there should be a comment pointing to https://github.com/CTSRD-CHERI/paper-revocation3/issues/28 and https://github.com/CTSRD-CHERI/paper-revocation3/issues/33 (which might want to move) as the right way forward?